security/pkg/credentialfetcher/plugin/gce.go

```diff
@@ -47,14 +63,63 @@ func CreateGCEPlugin(audience, jwtPath, identityProvider string) *GCEPlugin {
 		jwtPath:          jwtPath,
 		identityProvider: identityProvider,
 	}
+	go p.startTokenRotationJob()
 	return p
 }
 
+func (p *GCEPlugin) Stop() {
+	p.closing <- true
+}
+
+func (p *GCEPlugin) startTokenRotationJob() {
+	// Wake up once in a while and refresh GCE VM credential.
+	p.rotationTicker = time.NewTicker(rotationInterval)
+	for {
+		select {
+		case <-p.rotationTicker.C:
+			p.rotate()
+		case <-p.closing:
+			if p.rotationTicker != nil {
+				p.rotationTicker.Stop()
+			}
+		}
+	}
+}
+
+func (p *GCEPlugin) rotate() {
+	if p.shouldRotate(time.Now()) {
+		if _, err := p.GetPlatformCredential(); err != nil {
+			gcecredLog.Errorf("credential refresh failed: %+v", err)
+		}
+	}
+}
+
+func (p *GCEPlugin) shouldRotate(now time.Time) bool {
+	p.tokenMutex.RLock()
+	defer p.tokenMutex.RUnlock()
+
+	if p.tokenCache == "" {
+		return true
+	}
+	exp, err := util.GetExp(p.tokenCache)
+	// When fails to get expiration time from token, always refresh the token
+	if err != nil || exp.IsZero() {
```

Comment:
When exp.IsZero() is true, do you consider the token expired? It seems that the answer is "yes" here, but "false" in jwtUtill.go?

Reply:
exp.IsZero() means we don't get expiration time, for example, failure in decoding token, or missing `exp` field in the token. A valid expiration time will not be zero. `exp` stands for expiration time, not `expired`.

Reply:
It looks like the inconsistency is because:
(1) in the case of k8s first party token, missing "exp" claim is valid, which means that the token will never expire. This is the current implementation of IsJwtExpired(). 
(2) in the case of GCE identity token, missing "exp" claim is invalid. We always expect that "exp" claim exists. If "exp" is missing, the token is considered expired and should be refreshed. 

Should we update IsJwtExpired() to add a parameter "expRequired"? This will help distinguish the two cases, and clarify the hidden assumption.

Reply:
btw, I did a code search. It seems that IsJwtExpired() is not used anywhere expect the test.

Reply:
I will remove `IsJwtExpired()`. `shouldRotate()` is doing what we expect here, if it fails to get a valid expiration time for any reason, returns true. That will refetch the token.